File: Priority_Queue/378. Kth Smallest Element in a Sorted Matrix/378. Kth Smallest Element in a Sorted Matrix.cpp
class Solution {
public:
    int kthSmallest(vector<vector<int>>& matrix, int k) {
        int N = matrix.size();
        set<vector<int>>Set;
        Set.insert({matrix[0][0], 0, 0});
        
        int count = 0;
        while (count < k) {
            int val = (*Set.begin())[0];
            int x = (*Set.begin())[1];
            int y = (*Set.begin())[2];
            //pop first element
            Set.erase(Set.begin());
            count++;
            if (count == k) return val;
            
            if (y+1 < N)
                Set.insert({matrix[x][y+1], x, y+1});
            if (x+1 < N)
                Set.insert({matrix[x+1][y], x+1, y});
                 
        }
        
        return -1;
    }
};

File: Priority_Queue/378. Kth Smallest Element in a Sorted Matrix/378. Kth Smallest Element in a Sorted Matrix.java
class Solution {
    public int kthSmallest(int[][] matrix, int k) {
        int n = matrix.length;
        int left = matrix[0][0];
        int right = matrix[n-1][n-1];
        while (left < right) {
            int mid = left + (right-left) /2;
            
            if (smallerEquals(mid, matrix) < k) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    }
    private int smallerEquals(int val, int[][] matrix) {
        int n = matrix.length;
        int i = n-1, j = 0;
        int count = 0;
        while (i >= 0 && j < n) {
            if (matrix[i][j] > val) {
                i--;
            } else {
                count += (i+1);
                j++;
            }
        }
        return count;
    }
}
